File: typescript/Structural/12-Proxy/Real/JobCardRegistry.ts
import { IJobCardAccess } from "../Subject/IJobCardAccess";
import { JobCard } from "../Model/JobCard";

export class JobCardRegistry implements IJobCardAccess {
  private store = new Map<string, JobCard>();

  register(card: JobCard): void {
    this.store.set(card.jobId, card);
    console.log(`  [Registry] Registered: ${card}`);
  }

  getJobCard(jobId: string): JobCard | undefined {
    return this.store.get(jobId);
  }

  getAll(): ReadonlyArray<JobCard> {
    return Array.from(this.store.values());
  }
}

File: csharp/Structural/12-Proxy/Real/JobCardRegistry.cs
public class JobCardRegistry : IJobCardAccess
{
    private readonly Dictionary<string, JobCard> _store = new();

    public void Register(JobCard card)
    {
        _store[card.JobId] = card;
        Console.WriteLine($"  [Registry] Registered: {card}");
    }

    public JobCard? GetJobCard(string jobId)
    {
        _store.TryGetValue(jobId, out var card);
        return card;
    }

    public IReadOnlyList<JobCard> GetAll() =>
        _store.Values.ToList().AsReadOnly();
}
